// @cloudcare/browser-rum/cjs/domain/replay/deflate/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DeflateEncoderStreamId", {
  enumerable: true,
  get: function get() {
    return _deflateEncoder.DeflateEncoderStreamId;
  }
});
Object.defineProperty(exports, "DeflateWorkerStatus", {
  enumerable: true,
  get: function get() {
    return _deflateWorker.DeflateWorkerStatus;
  }
});
Object.defineProperty(exports, "createDeflateEncoder", {
  enumerable: true,
  get: function get() {
    return _deflateEncoder.createDeflateEncoder;
  }
});
Object.defineProperty(exports, "getDeflateWorkerStatus", {
  enumerable: true,
  get: function get() {
    return _deflateWorker.getDeflateWorkerStatus;
  }
});
Object.defineProperty(exports, "resetDeflateWorkerState", {
  enumerable: true,
  get: function get() {
    return _deflateWorker.resetDeflateWorkerState;
  }
});
Object.defineProperty(exports, "startDeflateWorker", {
  enumerable: true,
  get: function get() {
    return _deflateWorker.startDeflateWorker;
  }
});
var _deflateEncoder = require("./deflateEncoder");
var _deflateWorker = require("./deflateWorker");
