// @cloudcare/browser-rum/cjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PRIVACY_CLASS_MASK_USER_INPUT = exports.PRIVACY_CLASS_MASK = exports.PRIVACY_CLASS_HIDDEN = exports.PRIVACY_CLASS_ALLOW = exports.PRIVACY_ATTR_VALUE_MASK_USER_INPUT = exports.PRIVACY_ATTR_VALUE_MASK = exports.PRIVACY_ATTR_VALUE_HIDDEN = exports.PRIVACY_ATTR_VALUE_ALLOW = exports.PRIVACY_ATTR_NAME = exports.NodePrivacyLevel = exports.FORM_PRIVATE_TAG_NAMES = exports.CENSORED_STRING_MARK = exports.CENSORED_IMG_MARK = void 0;
var _browserCore = require("@cloudcare/browser-core");
var NodePrivacyLevel = {
  IGNORE: 'ignore',
  HIDDEN: 'hidden',
  ALLOW: _browserCore.DefaultPrivacyLevel.ALLOW,
  MASK: _browserCore.DefaultPrivacyLevel.MASK,
  MASK_USER_INPUT: _browserCore.DefaultPrivacyLevel.MASK_USER_INPUT
};
exports.NodePrivacyLevel = NodePrivacyLevel;
var PRIVACY_ATTR_NAME = 'data-gc-privacy';

// Privacy Attrs
exports.PRIVACY_ATTR_NAME = PRIVACY_ATTR_NAME;
var PRIVACY_ATTR_VALUE_ALLOW = 'allow';
exports.PRIVACY_ATTR_VALUE_ALLOW = PRIVACY_ATTR_VALUE_ALLOW;
var PRIVACY_ATTR_VALUE_MASK = 'mask';
exports.PRIVACY_ATTR_VALUE_MASK = PRIVACY_ATTR_VALUE_MASK;
var PRIVACY_ATTR_VALUE_MASK_USER_INPUT = 'mask-user-input';
exports.PRIVACY_ATTR_VALUE_MASK_USER_INPUT = PRIVACY_ATTR_VALUE_MASK_USER_INPUT;
var PRIVACY_ATTR_VALUE_HIDDEN = 'hidden';

// Privacy Classes - not all customers can set plain HTML attributes, so support classes too
exports.PRIVACY_ATTR_VALUE_HIDDEN = PRIVACY_ATTR_VALUE_HIDDEN;
var PRIVACY_CLASS_ALLOW = 'gc-privacy-allow';
exports.PRIVACY_CLASS_ALLOW = PRIVACY_CLASS_ALLOW;
var PRIVACY_CLASS_MASK = 'gc-privacy-mask';
exports.PRIVACY_CLASS_MASK = PRIVACY_CLASS_MASK;
var PRIVACY_CLASS_MASK_USER_INPUT = 'gc-privacy-mask-user-input';
exports.PRIVACY_CLASS_MASK_USER_INPUT = PRIVACY_CLASS_MASK_USER_INPUT;
var PRIVACY_CLASS_HIDDEN = 'gc-privacy-hidden';

// Private Replacement Templates
exports.PRIVACY_CLASS_HIDDEN = PRIVACY_CLASS_HIDDEN;
var CENSORED_STRING_MARK = '***';
exports.CENSORED_STRING_MARK = CENSORED_STRING_MARK;
var CENSORED_IMG_MARK = 'data:image/gif;base64,R0lGODlhAQABAIAAAMLCwgAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==';
exports.CENSORED_IMG_MARK = CENSORED_IMG_MARK;
var FORM_PRIVATE_TAG_NAMES = {
  INPUT: true,
  OUTPUT: true,
  TEXTAREA: true,
  SELECT: true,
  OPTION: true,
  DATALIST: true,
  OPTGROUP: true
};
exports.FORM_PRIVATE_TAG_NAMES = FORM_PRIVATE_TAG_NAMES;
