// @cloudcare/browser-rum/cjs/domain/tracing/jaegerTracer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.JaegerTracer = JaegerTracer;
var _traceIdentifier = require("./traceIdentifier");
/**
 *
 * @param {*} configuration  配置信息
 */
function JaegerTracer(configuration, traceSampled) {
  this._traceId = new _traceIdentifier.TraceIdentifier();
  this._spanId = new _traceIdentifier.TraceIdentifier();
  this._traceSampled = traceSampled;
  this.is128Bit = configuration.traceId128Bit;
}
JaegerTracer.prototype = {
  isTracingSupported: function isTracingSupported() {
    return (0, _traceIdentifier.getCrypto)() !== undefined;
  },
  getSpanId: function getSpanId() {
    return this._spanId.toPaddedHexadecimalString();
  },
  getTraceId: function getTraceId() {
    return this.is128Bit ? '0000000000000000' + this._traceId.toPaddedHexadecimalString() : this._traceId.toPaddedHexadecimalString();
  },
  getUberTraceId: function getUberTraceId() {
    //{trace-id}:{span-id}:{parent-span-id}:{flags}
    return this.getTraceId() + ':' + this.getSpanId() + ':' + '0' + ':' + (this._traceSampled ? '1' : '0');
  },
  makeTracingHeaders: function makeTracingHeaders() {
    return {
      'uber-trace-id': this.getUberTraceId()
    };
  }
};
