// @cloudcare/browser-rum/cjs/domain/tracing/zipkinSingleTracer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZipkinSingleTracer = ZipkinSingleTracer;
var _traceIdentifier = require("./traceIdentifier");
/**
 *
 * @param {*} configuration  配置信息
 */
function ZipkinSingleTracer(configuration, traceSampled) {
  this._traceId = new _traceIdentifier.TraceIdentifier();
  this._spanId = new _traceIdentifier.TraceIdentifier();
  this._traceSampled = traceSampled;
}
ZipkinSingleTracer.prototype = {
  isTracingSupported: function isTracingSupported() {
    return (0, _traceIdentifier.getCrypto)() !== undefined;
  },
  getSpanId: function getSpanId() {
    return this._spanId.toPaddedHexadecimalString();
  },
  getTraceId: function getTraceId() {
    return this._traceId.toPaddedHexadecimalString();
  },
  getB3Str: function getB3Str() {
    //{TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
    return this.getTraceId() + '-' + this.getSpanId() + '-' + (this._traceSampled ? '1' : '0');
  },
  makeTracingHeaders: function makeTracingHeaders() {
    return {
      b3: this.getB3Str()
    };
  }
};
