// @cloudcare/browser-rum/cjs/domain/rumEventsCollection/longTask/longTaskCollection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startLongTaskCollection = startLongTaskCollection;
var _browserCore = require("@cloudcare/browser-core");
function startLongTaskCollection(lifeCycle, sessionManager) {
  lifeCycle.subscribe(_browserCore.LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED, function (entries) {
    for (var i = 0; i < entries.length; i++) {
      var entry = entries[i];
      if (entry.entryType !== 'longtask') {
        return;
      }
      var session = sessionManager.findTrackedSession(entry.startTime);
      if (!session) {
        break;
      }
      var startClocks = (0, _browserCore.relativeToClocks)(entry.startTime);
      var rawRumEvent = {
        date: startClocks.timeStamp,
        long_task: {
          id: (0, _browserCore.UUID)(),
          duration: (0, _browserCore.toServerDuration)(entry.duration)
        },
        type: _browserCore.RumEventType.LONG_TASK
      };
      lifeCycle.notify(_browserCore.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, {
        rawRumEvent: rawRumEvent,
        startTime: startClocks.relative,
        domainContext: {
          performanceEntry: entry.toJSON()
        }
      });
    }
  });
}
