// @cloudcare/browser-core/cjs/browser/cookie.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.COOKIE_ACCESS_DELAY = void 0;
exports.areCookiesAuthorized = areCookiesAuthorized;
exports.deleteCookie = deleteCookie;
exports.getCookie = getCookie;
exports.getCurrentSite = getCurrentSite;
exports.getInitCookie = getInitCookie;
exports.resetInitCookies = resetInitCookies;
exports.setCookie = setCookie;
var _tools = require("../helper/tools");
var COOKIE_ACCESS_DELAY = _tools.ONE_SECOND;
exports.COOKIE_ACCESS_DELAY = COOKIE_ACCESS_DELAY;
function setCookie(name, value, expireDelay, options) {
  var date = new Date();
  date.setTime(date.getTime() + expireDelay);
  var expires = 'expires=' + date.toUTCString();
  var sameSite = options && options.crossSite ? 'none' : 'strict';
  var domain = options && options.domain ? ';domain=' + options.domain : '';
  var secure = options && options.secure ? ';secure' : '';
  var partitioned = options && options.partitioned ? ';partitioned' : '';
  document.cookie = name + '=' + value + ';' + expires + ';path=/;samesite=' + sameSite + domain + secure + partitioned;
}
function getCookie(name) {
  return (0, _tools.findCommaSeparatedValue)(document.cookie, name);
}
var initCookieParsed;
/**
 * Returns a cached value of the cookie. Use this during SDK initialization (and whenever possible)
 * to avoid accessing document.cookie multiple times.
 */
function getInitCookie(name) {
  if (!initCookieParsed) {
    initCookieParsed = (0, _tools.findCommaSeparatedValues)(document.cookie);
  }
  return initCookieParsed.get(name);
}
function resetInitCookies() {
  initCookieParsed = undefined;
}
function deleteCookie(name, options) {
  setCookie(name, '', 0, options);
}
function areCookiesAuthorized(options) {
  if (document.cookie === undefined || document.cookie === null) {
    return false;
  }
  try {
    // Use a unique cookie name to avoid issues when the SDK is initialized multiple times during
    // the test cookie lifetime
    var testCookieName = "gc_cookie_test_".concat((0, _tools.UUID)());
    var testCookieValue = 'test';
    setCookie(testCookieName, testCookieValue, _tools.ONE_SECOND, options);
    return getCookie(testCookieName) === testCookieValue;
  } catch (error) {
    return false;
  }
}

/**
 * No API to retrieve it, number of levels for subdomain and suffix are unknown
 * strategy: find the minimal domain on which cookies are allowed to be set
 * https://web.dev/same-site-same-origin/#site
 */
var getCurrentSiteCache;
function getCurrentSite() {
  if (getCurrentSiteCache === undefined) {
    // Use a unique cookie name to avoid issues when the SDK is initialized multiple times during
    // the test cookie lifetime
    var testCookieName = "gc_site_test_".concat((0, _tools.UUID)());
    var testCookieValue = 'test';
    var domainLevels = window.location.hostname.split('.');
    var candidateDomain = domainLevels.pop();
    while (domainLevels.length && !getCookie(testCookieName)) {
      candidateDomain = domainLevels.pop() + '.' + candidateDomain;
      setCookie(testCookieName, testCookieValue, _tools.ONE_SECOND, {
        domain: candidateDomain
      });
    }
    getCurrentSiteCache = candidateDomain;
  }
  return getCurrentSiteCache;
}
