// @cloudcare/browser-rum/cjs/domain/trackEventCounts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trackEventCounts = trackEventCounts;
var _browserCore = require("@cloudcare/browser-core");
function trackEventCounts(data) {
  var lifeCycle = data.lifeCycle;
  var isChildEvent = data.isChildEvent;
  var callback = data.onChange;
  if (callback === undefined) {
    callback = _browserCore.noop;
  }
  var eventCounts = {
    errorCount: 0,
    longTaskCount: 0,
    resourceCount: 0,
    actionCount: 0,
    frustrationCount: 0
  };
  var subscription = lifeCycle.subscribe(_browserCore.LifeCycleEventType.RUM_EVENT_COLLECTED, function (event) {
    if (event.type === _browserCore.RumEventType.VIEW || !isChildEvent(event)) {
      return;
    }
    switch (event.type) {
      case _browserCore.RumEventType.ERROR:
        eventCounts.errorCount += 1;
        callback();
        break;
      case _browserCore.RumEventType.ACTION:
        if (event.action.frustration) {
          eventCounts.frustrationCount += event.action.frustration.type.length;
        }
        eventCounts.actionCount += 1;
        callback();
        break;
      case _browserCore.RumEventType.LONG_TASK:
        eventCounts.longTaskCount += 1;
        callback();
        break;
      case _browserCore.RumEventType.RESOURCE:
        eventCounts.resourceCount += 1;
        callback();
        break;
    }
  });
  return {
    stop: function stop() {
      subscription.unsubscribe();
    },
    eventCounts: eventCounts
  };
}
