// @cloudcare/browser-core/cjs/helper/observable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Observable = void 0;
exports.mergeObservables = mergeObservables;
var _tools = require("./tools");
var _Observable = function _Observable(onFirstSubscribe) {
  this.observers = [];
  this.onLastUnsubscribe = undefined;
  this.onFirstSubscribe = onFirstSubscribe;
};
_Observable.prototype = {
  subscribe: function subscribe(f) {
    if (!this.observers.length && this.onFirstSubscribe) {
      this.onLastUnsubscribe = this.onFirstSubscribe(this) || undefined;
    }
    this.observers.push(f);
    var _this = this;
    return {
      unsubscribe: function unsubscribe() {
        _this.observers = (0, _tools.filter)(_this.observers, function (other) {
          return f !== other;
        });
        if (!_this.observers.length && _this.onLastUnsubscribe) {
          _this.onLastUnsubscribe();
        }
      }
    };
  },
  notify: function notify(data) {
    (0, _tools.each)(this.observers, function (observer) {
      observer(data);
    });
  }
};
var Observable = _Observable;
exports.Observable = Observable;
function mergeObservables() {
  var observables = [].slice.call(arguments);
  return new Observable(function (globalObservable) {
    var subscriptions = (0, _tools.map)(observables, function (observable) {
      return observable.subscribe(function (data) {
        return globalObservable.notify(data);
      });
    });
    return function () {
      return (0, _tools.each)(subscriptions, function (subscription) {
        return subscription.unsubscribe();
      });
    };
  });
}
