// @cloudcare/browser-rum/cjs/domain/contexts/viewContexts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VIEW_CONTEXT_TIME_OUT_DELAY = void 0;
exports.startViewContexts = startViewContexts;
var _browserCore = require("@cloudcare/browser-core");
var VIEW_CONTEXT_TIME_OUT_DELAY = _browserCore.SESSION_TIME_OUT_DELAY;
exports.VIEW_CONTEXT_TIME_OUT_DELAY = VIEW_CONTEXT_TIME_OUT_DELAY;
function startViewContexts(lifeCycle) {
  var viewContextHistory = new _browserCore.ContextHistory(VIEW_CONTEXT_TIME_OUT_DELAY);
  lifeCycle.subscribe(_browserCore.LifeCycleEventType.BEFORE_VIEW_CREATED, function (view) {
    viewContextHistory.add(buildViewContext(view), view.startClocks.relative);
  });
  lifeCycle.subscribe(_browserCore.LifeCycleEventType.AFTER_VIEW_ENDED, function (data) {
    viewContextHistory.closeActive(data.endClocks.relative);
  });
  lifeCycle.subscribe(_browserCore.LifeCycleEventType.SESSION_RENEWED, function () {
    viewContextHistory.reset();
  });
  function buildViewContext(view) {
    return {
      service: view.service,
      version: view.version,
      id: view.id,
      name: view.name
    };
  }
  return {
    findView: function findView(startTime) {
      return viewContextHistory.find(startTime);
    },
    stop: function stop() {
      viewContextHistory.stop();
    }
  };
}
