// @cloudcare/browser-core/cjs/transport/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Batch", {
  enumerable: true,
  get: function get() {
    return _batch.Batch;
  }
});
Object.defineProperty(exports, "canUseEventBridge", {
  enumerable: true,
  get: function get() {
    return _eventBridge.canUseEventBridge;
  }
});
Object.defineProperty(exports, "createFlushController", {
  enumerable: true,
  get: function get() {
    return _flushController.createFlushController;
  }
});
Object.defineProperty(exports, "createHttpRequest", {
  enumerable: true,
  get: function get() {
    return _httpRequest.createHttpRequest;
  }
});
Object.defineProperty(exports, "getEventBridge", {
  enumerable: true,
  get: function get() {
    return _eventBridge.getEventBridge;
  }
});
Object.defineProperty(exports, "processedMessageByDataMap", {
  enumerable: true,
  get: function get() {
    return _batch.processedMessageByDataMap;
  }
});
Object.defineProperty(exports, "startBatchWithReplica", {
  enumerable: true,
  get: function get() {
    return _startBatchWithReplica.startBatchWithReplica;
  }
});
var _httpRequest = require("./httpRequest");
var _batch = require("./batch");
var _startBatchWithReplica = require("./startBatchWithReplica");
var _flushController = require("./flushController");
var _eventBridge = require("./eventBridge");
