// @cloudcare/browser-core/cjs/transport/startBatchWithReplica.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startBatchWithReplica = startBatchWithReplica;
var _batch = require("./batch");
var _httpRequest = require("./httpRequest");
var _flushController = require("./flushController");
function startBatchWithReplica(configuration, endpointUrl, reportError, pageExitObservable, sessionExpireObservable) {
  var primaryBatch = createBatch(endpointUrl);
  function createBatch(endpointUrl) {
    return new _batch.Batch((0, _httpRequest.createHttpRequest)(endpointUrl, configuration.batchBytesLimit, configuration.sendContentTypeByJson, reportError), (0, _flushController.createFlushController)({
      messagesLimit: configuration.batchMessagesLimit,
      bytesLimit: configuration.batchBytesLimit,
      durationLimit: configuration.flushTimeout,
      pageExitObservable: pageExitObservable,
      sessionExpireObservable: sessionExpireObservable
    }), configuration.messageBytesLimit, configuration.sendContentTypeByJson);
  }
  return {
    add: function add(message) {
      primaryBatch.add(message);
    }
  };
}
