// @cloudcare/browser-core/cjs/synthetics/syntheticsWorkerValues.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYNTHETICS_TEST_ID_COOKIE_NAME = exports.SYNTHETICS_RESULT_ID_COOKIE_NAME = exports.SYNTHETICS_INJECTS_RUM_COOKIE_NAME = void 0;
exports.getSyntheticsResultId = getSyntheticsResultId;
exports.getSyntheticsTestId = getSyntheticsTestId;
exports.willSyntheticsInjectRum = willSyntheticsInjectRum;
var _cookie = require("../browser/cookie");
var SYNTHETICS_TEST_ID_COOKIE_NAME = 'guance-synthetics-public-id';
exports.SYNTHETICS_TEST_ID_COOKIE_NAME = SYNTHETICS_TEST_ID_COOKIE_NAME;
var SYNTHETICS_RESULT_ID_COOKIE_NAME = 'guance-synthetics-result-id';
exports.SYNTHETICS_RESULT_ID_COOKIE_NAME = SYNTHETICS_RESULT_ID_COOKIE_NAME;
var SYNTHETICS_INJECTS_RUM_COOKIE_NAME = 'guance-synthetics-injects-rum';
exports.SYNTHETICS_INJECTS_RUM_COOKIE_NAME = SYNTHETICS_INJECTS_RUM_COOKIE_NAME;
function willSyntheticsInjectRum() {
  return Boolean(window._GUANCE_SYNTHETICS_INJECTS_RUM || (0, _cookie.getCookie)(SYNTHETICS_INJECTS_RUM_COOKIE_NAME));
}
function getSyntheticsTestId() {
  var value = window._GUANCE_SYNTHETICS_PUBLIC_ID || (0, _cookie.getCookie)(SYNTHETICS_TEST_ID_COOKIE_NAME);
  return typeof value === 'string' ? value : undefined;
}
function getSyntheticsResultId() {
  var value = window._GUANCE_SYNTHETICS_RESULT_ID || (0, _cookie.getCookie)(SYNTHETICS_RESULT_ID_COOKIE_NAME);
  return typeof value === 'string' ? value : undefined;
}
